<a href="https://colab.research.google.com/github/olegpant/python/blob/main/%D0%9A%D0%BE%D0%BF%D0%B8%D1%8F_%D0%B1%D0%BB%D0%BE%D0%BA%D0%BD%D0%BE%D1%82%D0%B0_%22hw_10_4_apply%2C_groupby%2C_pivot_table_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import pandas as pd

In [3]:
df = pd.read_csv('/content/drive/MyDrive/DataAnalystEdc/Pandas/data/supermarket_sales.csv')
df.head()

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3


1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [4]:
df.groupby('Branch')['Total'].sum().sort_values(ascending=False)

,Total
Branch,
C,110568.7065
A,106200.3705
B,106197.6720


2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [5]:
df.groupby(['Branch','City'])['Total'].sum().sort_values(ascending=False)

,,Total
Branch,City,
C,Naypyitaw,110568.7065
A,Yangon,106200.3705
B,Mandalay,106197.6720


З найвищим прибутком філіал знайходиться в місті Naypyitaw.

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [6]:
invoices_by_category = df.pivot_table(index='Branch', columns='Product line', values='Invoice ID', aggfunc='count')
invoices_by_category

Product line,Electronic accessories,Fashion accessories,Food and beverages,Health and beauty,Home and lifestyle,Sports and travel
Branch,,,,,,
A,60,51,58,47,65,59
B,55,62,50,53,50,62
C,55,65,66,52,45,45


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [7]:
invoices_by_category['Electronic accessories'].idxmax()

'A'

4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


1.Gender x Product line. Порівняти, які категорії найчастіше купують чоловіки та жінки.Quantity, кількість чеків та середній чек. Інакше велика сума може пояснюватися просто дорожчими товарами.

2.Payment x Total. Чи повʼязаний спосіб оплати з розміром чека.

3.Member vs Normal. Порівняти, скільки виручки та gross income приносять Member і Normal, але обов’язково поруч подивитися їхню кількість. Cередній чек, середню кількість товарів у чеку та улюблені категорії. Це покаже, чи Member дійсно цінніші для бізнесу, а не просто становлять більшу частку клієнтів.

In [37]:
# Gender × Product line.
df.groupby(['Gender','Product line']).agg(
    count_invoice = ('Invoice ID', 'count'),
    avg_total_productLine = ('Total', 'mean'),
    total_quantity = ('Quantity', 'sum')
)

count_invoice  avg_total_productLine  \
Gender Product line                                                   
Female Electronic accessories             84             322.643125   
       Fashion accessories                96             317.056250   
       Food and beverages                 90             368.565750   
       Health and beauty                  64             290.015414   
       Home and lifestyle                 79             380.213639   
       Sports and travel                  88             324.712739   
Male   Electronic accessories             86             316.691965   
       Fashion accessories                82             291.079207   
       Food and beverages                 84             273.499125   
       Health and beauty                  88             348.099460   
       Home and lifestyle                 81             294.136241   
       Sports and travel                  78             340.360327   

                               total_quantity  
Gender Product line                            
Female Electronic accessories             488  
       Fashion accessories                530  
       Food and beverages                 514  
       Health and beauty                  343  
       Home and lifestyle                 498  
       Sports and travel                  496  
Male   Electronic accessories             483  
       Fashion accessories                372  
       Food and beverages                 438  
       Health and beauty                  511  
       Home and lifestyle                 413  
       Sports and travel                  424

Перше, що кидається в очі, — Fashion accessories. У жінок кількість чеків вища: 96 проти 82 у чоловіків. Середній чек також вищий — близько 317 проти 291, а загальна кількість придбаних товарів становить 530 проти 372. Це свідчить про те, що жінки в цій категорії не тільки купують частіше, але й у середньому витрачають більше за одну покупку. Схожа ситуація спостерігається в категорії «Food & Beverage». У жінок середній чек значно вищий — близько 368 проти 273 у чоловіків, а загальна кількість придбаних товарів становить 514 проти 438. Кількість чеків при цьому порівнянна. Це свідчить про те, що жінки в цій категорії роблять більш великі покупки.Зворотна тенденція спостерігається в категорії Health & Beauty. Тут чоловіки активніші: кількість чеків становить 88 проти 64 у жінок, середній чек — близько 348 проти 290, а кількість придбаних товарів — 511 проти 343. Тобто в цій категорії чоловіки купують частіше і витрачають більше.

Для Fashion Accessories можна запропонувати посилити чоловічий сегмент: розширити асортимент товарів для чоловіків, протестувати окремі промо-викладки та акції, орієнтовані на чоловічу аудиторію. Водночас жіночий сегмент уже показує вищу активність за кількістю покупок, тому для нього можна використовувати таргетовані акції, набори товарів або крос-продажі, щоб додатково збільшити середній чек.
Для Food & Beverage доцільно використати високий потенціал жіночої аудиторії через персоналізовані акції, комплекти товарів і пропозиції на категорії, які вони купують найчастіше.
Для Health & Beauty варто окремо проаналізувати, які саме товари краще купують чоловіки, і перевірити, чи достатньо представлений асортимент для жінок. Якщо виявиться, що жіночий сегмент недоотримує потрібні продукти, можна розширити асортимент, додати цільові промоакції та покращити викладку товарів для збільшення продажів у цій категорії.

In [38]:
# 2.Payment x Total
df.groupby(['Payment']).agg(
    count_invoice_by_payment = ('Invoice ID', 'count'),
    sum_total = ('Total','sum'),
    avg_total_payment = ('Total', 'mean'))

,count_invoice_by_payment,sum_total,avg_total_payment
Payment,,,
Cash,344,112206.570,326.181890
Credit card,311,100767.072,324.009878
Ewallet,345,109993.107,318.820600


Згідно з таблицею, кількість інвойсів за різними способами оплати є приблизно однаковою, понад 300 для кожної категорії. Середній чек також майже не відрізняється й становить близько 300 незалежно від способу оплати. Отже, спосіб оплати не має помітного впливу на розмір середнього чека. Значної залежності між payment method і сумою покупки в цих даних не спостерігається.
Cпосіб оплати не є фактором, який помітно впливає на середній чек. Тому немає сенсу інвестувати в просування конкретного payment method саме з метою збільшення суми покупки. Для цього краще використовувати інші важелі: промоакції, крос-продажі, роботу з асортиментом і продуктовими категоріями.
Водночас усі способи оплати використовуються приблизно однаково часто. Тому бізнесу важливо підтримувати всі доступні канали оплати: картки, готівку та електронні гаманці. Щоб не створювати незручностей для значної частини покупців і не втрачати потенційні продажі.

In [40]:
# 3.Member vs Norma
df.groupby(['Customer type']).agg(
    count_cutomerType = ('Customer type', 'count'),
    sum_total_by_cutomerType = ('Total', 'sum'),
    sum_gross_income = ('gross income', 'sum'),
    avg_quantity = ('Quantity', 'mean'),
    avg_total_by_cutomerType = ('Total', 'mean')

)

,count_cutomerType,sum_total_by_cutomerType,sum_gross_income,avg_quantity,avg_total_by_cutomerType
Customer type,,,,,
Member,501,164223.444,7820.164,5.558882,327.791305
Normal,499,158743.305,7559.205,5.460922,318.122856


Кількість Member і звичайних клієнтів майже однакова: 501 проти 499. При цьому Member приносять трохи більше виручки, переважно завдяки вищому середньому чеку: 327 проти 318, а не через більшу кількість клієнтів.
Отже, Member є дещо ціннішим сегментом, але різниця поки невелика. Для бізнесу варто перевірити, чи можна посилити ефект програми лояльності через персональні пропозиції, бонуси, спеціальні акції та крос-продажі, щоб збільшити середній чек і частоту повторних покупок.
Водночас важливо не робити надто сильний висновок лише з цієї різниці: спочатку варто додатково порівняти частоту покупок, структуру категорій і витрати на одного клієнта.